clang/test/CodeGenCXX/cheri/single-cap-no-byval.cpp

```diff
@@ -0,0 +1,88 @@
+// NOTE: Assertions have been autogenerated by utils/update_cc_test_checks.py UTC_ARGS: --function-signature
+// Check that the generated IR for this code does not add a byval attribute to the call argument
+// Previous versions of Clang would incorrectly add a byval attribute that was then ignored in the backend:
+// declare inreg { i8 addrspace(200)* } @_ZN3JSC2VM14throwExceptionEPNS_9ExecStateENS_7JSValueE(%"class.JSC::VM" addrspace(200)*, %"class.JSC::ExecState" addrspace(200)*, %"class.JSC::JSValue" addrspace(200)* align 16) #2
+// The backend no longer includes the hack to ignore byval, so check that passing by value does not add the byval attribute
+
+// RUN: %cheri_purecap_cc1 -O2 -emit-llvm -o - %s | FileCheck %s --check-prefixes CHECK,MIPS
+// RUN: %riscv64_cheri_purecap_cc1 -O2 -emit-llvm -o - %s | FileCheck %s --check-prefixes CHECK,RV64
+
+union JsValue {
+  long l;
+  __uintcap_t x;
+};
+struct ExecState;
+
+JsValue throwException(ExecState *, JsValue);
+
+// CHECK-LABEL: define {{[^@]+}}@_Z4testv()
+// MIPS: call inreg i8 addrspace(200)* @_Z14throwExceptionP9ExecState7JsValue(%struct.ExecState addrspace(200)* null, i8 addrspace(200)* inreg %{{[a-z.0-9]+}})
+// RV64: call %union.JsValue @_Z14throwExceptionP9ExecState7JsValue(%struct.ExecState addrspace(200)* null, i8 addrspace(200)* %{{[a-z.0-9]+}})
+//
+// MIPS: declare inreg i8 addrspace(200)* @_Z14throwExceptionP9ExecState7JsValue(%struct.ExecState addrspace(200)*, i8 addrspace(200)* inreg)
+// RV64: declare %union.JsValue @_Z14throwExceptionP9ExecState7JsValue(%struct.ExecState addrspace(200)*, i8 addrspace(200)*)
```

Comment:
@jrtc27 Is it intentional that RISC-V changes the argument type to `i8 addrspace(200)*` but not the return type?

Reply:
Hm, they're supposed to be handled identically, save for C++ special cases (which shouldn't apply here, and I checked this happens even for C). The only difference _should_ be how many GPRs/FPRs it claims are available (2/{0,2} vs 8/{0,8}). I wonder if this particular conversion is happening somewhere else other than RISCVABIInfo?

Reply:
Yeah, RISCVABIInfo::classifyArgumentType sees them both as the union and does:
```cxx
  if (IsSingleCapStruct)
    return ABIArgInfo::getDirect();
```

Reply:
The only thing I can see that could break this is
```
  if (!getCXXABI().classifyReturnType(FI))
    FI.getReturnInfo() = classifyReturnType(RetTy);
```
Let's see what the dbugger says.

Reply:
I wouldn't see `classifyArgumentType` being called twice in that case, and printf debugging says that condition is always true (ie `FI.getReturnInfo() = classifyReturnType(RetTy);` _does_ get called as we expect).

Reply:
Yeah looks like it's happening in CGCall.cpp. I guess for MIPS the ABI info already does the return type flattening.

Reply:
Generate assembly looks sensible so this difference shouldn't matter.